File: miscellaneous/matrix-bfs-traversal/20221123/matrix_bfs.py
from typing import List


def matrix_bfs(matrix: List[List[int]]) -> List[int]:
    values: List[int] = []
    queue = [[0, 0]]
    DIRECTIONS = (-1, 0), (0, 1), (1, 0), (0, -1)
    seen = [[False for x in range(len(matrix[0]))] for y in range(len(matrix))]
    seen[0][0] = True

    while len(queue):
        [row, col] = queue.pop(0)
        values.append(matrix[row][col])

        for [row_adj, col_adj] in DIRECTIONS:
            new_row = row + row_adj
            new_col = col + col_adj

            if in_bounds(new_row, new_col, matrix) and not seen[new_row][new_col]:
                queue.append([new_row, new_col])
                seen[new_row][new_col] = True

    return values


def in_bounds(row: int, col: int, matrix: List[List[int]]) -> bool:
    return 0 <= row < len(matrix) and 0 <= col < len(matrix[0])

File: miscellaneous/matrix-bfs-traversal/20230107/matrix-bfs.js
'use strict';

const moves = [[-1, 0], [0, 1], [1, 0], [0, -1]];

function matrixBfs(matrix) {
    const values = [];
    const queue = [[0, 0]];
    const seen = [...Array(matrix.length)]
        .map(() => Array(matrix[0].length).fill(false));
    seen[0][0] = true;

    while (queue.length) {
        const [row, col] = queue.shift();
        values.push(matrix[row][col])

        for (const [moveY, moveX] of moves) {
            const newRow = row + moveY;
            const newCol = col + moveX;

            if (inBounds(newRow, newCol, matrix) && !seen[newRow][newCol]) {
                queue.push([newRow, newCol]);
                seen[newRow][newCol] = true;
            }
        }
    }

    return values;
}

function inBounds(row, col, matrix) {
    return row >= 0 && row < matrix.length && col >= 0 && col < matrix[0].length;
}

module.exports = { matrixBfs };